pkg/manifestclient/write_roundtripper.go

```diff
@@ -213,8 +216,14 @@ func (mrt *writeTrackingRoundTripper) roundTrip(req *http.Request) ([]byte, erro
 	ret := &unstructured.Unstructured{Object: map[string]interface{}{}}
 	ret.SetName(serializedRequest.ActionMetadata.Name)
 	ret.SetNamespace(serializedRequest.ActionMetadata.Namespace)
-	if actionHasRuntimeObjectBody { // TODO might be able to do something generally based on discovery if absolutely necessary
+	if actionHasRuntimeObjectBody {
 		ret.SetGroupVersionKind(bodyObj.GetObjectKind().GroupVersionKind())
+	} else {
+		kindForResource, err := mrt.getKindForResourceFunc(gvr)
```

Comment:
right here, we should be able to use discovery to determine the GVK, same as we do for reading the list kind.